### ISSUE TITLE
kv: rate limiting and usage quotas for tenants

### ISSUE THREAD

**Issue by tbg** [cockroachdb/cockroach#47906]
See [design doc](https://docs.google.com/document/d/1EI8T-3N77g0OOlALwzRC0ddji1_4rH2xZvG9GbAa6D0/edit?ts=5e7a83b4#heading=h.4vc8e8rw3cvq)

Requests on behalf of tenants must be rate limited. Initially, this system should be lightweight in the sense that one limit applies to all tenants (or that exceptions are made manually) though we do expect it to evolve post phase 2. We need to figure out what to rate limit on and how, and we need to experimentally determine suitable limits for the envisioned deployments.

We need to decide whether usage quotas are to be applied at the KV layer (alternative: have deployment look at the metrics and shut down tenants that are above quota instead) and, if so, come up with a design.

**Comment by ajwerner** [comment]
Just turning my attention to this now. My sense is that this is a deeper issue which will require a number of iterations to get right.

In the first pass we have some luxuries which won't be available later. In the first iteration, there is like to be just a single client talking to a single leaseholder. I'm going to take this to mean that it's acceptable for rate limits to be applied on a per kv server basis with no distributed rate limiting. 

Even with the assumption that we only want to rate-limit on each KV server, what should we be rate limiting? What knobs should we permit? How should we implement the rate limit? Ideally we'd rate-limit based on a combination of byte-rate and request-rate. The normal way this is done is with a token bucket. That only works perfectly when you know the resource being allocated a prior, which is not something we definitely know here. In particular, if we want to enforce byte rates, we could assume that a request will read the number of bytes that it sets as its maximum but that seems pretty crazy. A batch with a bunch of `Get` requests also don't have the same byte limit semantics.

I think it makes sense to do something like a token bucket on byte rate that we allocate from using the limit and then replenish the delta based on the bytes actually read. We'll also want to limit QPS. In short, I don't think a simple wrapped `rate.Limiter` is going to do exactly what we need.

Let me think on how to do this best. I think the first step will be to type up a set of requirements and some interfaces then look towards implementation. My first instinct is that this might fit nicely into `quotapool` abstraction whereby each tenant maintains two dimensions of token buckets as the resource underneath, the general requests one and then a byte-oriented token bucket. We can have a single goroutine for all tenants tickle the token buckets to progress time as needed.

**Comment by tbg** [comment]
>  I'm going to take this to mean that it's acceptable for rate limits to be applied on a per kv server basis with no distributed rate limiting.

That's my understanding as well. In later stages (i.e. not initially) we will want some sort of distributed rate limiting (nodes exchanging smoothed-out averages of per-tenant usage which they can use to add an assumed base consumption? Definitely should not try to solve this now).

> Even with the assumption that we only want to rate-limit on each KV server, what should we be rate limiting?

We will initially want incoming bytes, incoming puts, outgoing bytes (note that DeleteRange would be throttled only via OutgoingBytes, see below for TargetBytes). In standard operation I would expect outgoing bytes to be the most important one.

You mention (implicitly) that request counts (and anything else that you can glean from incoming request) are handled by standard token buckets. For byte rates (and anything else like them), I was thinking the following:

1. require that there is a reasonable max consumption per request (for example, error out if `TargetBytes > 1mb` or unset).
2. apply rate limiting to responses instead of requests. For example for a scan with MaxBytes=1mb that ends up scanning 12kb, we take (and possibly block for some time) 12kb worth of tokens from the bucket before returning the response to the client.

This side-steps the problem at the cost of possibly allowing a burst of concurrent requests to execute concurrently when they'd normally be throttled to go one-by-one. I think this is OK - the request rate limiter attenuates this problem to some extent - with a req rate limit of 5/s and 1mb/s bytes rate limiter at worst the tenant can get 5mb/s out of the system (assuming requests finish ~instantly, we can guard against the case in which they don't separately). I think that could be preferable to the "subtract the max possible now and then release the delta back in" since it would allow us to use fairly tight rate limits (the other option seems to require overprovisioning, and possibly a lot of it)

> We can have a single goroutine for all tenants tickle the token buckets to progress time as needed.

I didn't think we would need this, isn't the way this is usually done by having the bucket retain a time of last access and moving time with each new access? I was imagining a `sync.Map` keyed on the tenantID and periodically (on long timescales, minutes, perhaps more) we go through and purge tenants that have not seen any recent activity).

The other thing you should think about is the interaction with metrics. It would be good if the rate limiters automatically registered with metrics in a way that exposes them via prometheus and to have some more general infra for that (i.e. the beginnings of #49825). Also I think (not sure it's really required though, depends on number of tenants and also tenant ID churn) there's room for an abstraction that "expires" per-tenant state across the various subsystems when we haven't seen that tenant used for a while (similar to what I'm describing for the rate limiter).

**Comment by ajwerner** [comment]
>  1. require that there is a reasonable max consumption per request (for example, error out if TargetBytes > 1mb or unset).

How does this work for `GetRequest`? I'll run with the suggestion for 2. and see what falls out. 

> I didn't think we would need this, isn't the way this is usually done by having the bucket retain a time of last access and moving time with each new access? I was imagining a sync.Map keyed on the tenantID and periodically (on long timescales, minutes, perhaps more) we go through and purge tenants that have not seen any recent activity).

Right, fair, I went and did some research on implementations after typing that.

**Comment by tbg** [comment]
> How does this work for GetRequest? I'll run with the suggestion for 2. and see what falls out.

We have no request limit for single KV pairs (similar to how a `ScanRequest` with `TargetBytes=1mb` will always return at least one row, even if that row is 100mb large), so I guess the answer is just that it won't be perfect in that case. But note that we would empty the bucket the moment we see the 10mb row, and would then drain the bucket for as long as it takes to get 10mb of quota. There might be some work to be done if we want fairness here as I think we do (we don't want a large returning request to be starved by new incoming requests).

The other approach which is maybe better is to thread rate limiting down with the context. At some point near the API surface you have

```go
ctx = something.WithRateLimiter(ctx, rateLimiter)
```

and then, maybe deep down in mvccScanner or even pebble,

```go
func (i *iterator) advance() bool {
  nextSize := i.nextSize()
  bytesbucket.WaitN(ctx, nextSize)
  return i.emit(i.readNext())
}
```

(this is totally fantasy code btw)

This context-based approach has the advantage that it can limit on virtually any resource, as that resource is acquired, at least as long as we stay in Go, which notably includes batch evaluation.

A similar approach appears promising for metrics as well.

**Comment by ajwerner** [comment]
I’ve been chewing on this throughout the day and what I’ve come up with is that I like the idea of applying the byte read rate limit after the fact. The primary problem with that approach is that a burst of requests could pile up and then be blocked after having served the request. We don’t really want that to happen too much so what I’m thinking is that in addition to the rate limit on qps we can have a concurrency limit as well. If requests are waiting for their tokens for the bytes they read, they’ll hold on to their quota for the concurrency budget.

**Comment by tbg** [comment]
> they’ll hold on to their quota for the concurrency budget.

That improves the situation for this particular problem, but how would we tune the concurrency limit? It seems hard to get right. It's totally reasonable for a user to send 100 concurrent requests if they all basically read ~nothing, so the size rate limit won't kick in. OTOH 100 concurrent requests that eat 1mb each is way too much.

It might be more targeted to make it such that incoming requests acquire a "courtesy byte" from the bucket before eval (after all, each request surely causes 1 byte in outgoing traffic no matter what). This has the nice effect that when a bunch of concurrent requests have snuck in and are now sucking the bucket empty on their way out, new incoming requests will block (I am assuming fairness here - outgoing requests get tokens before incoming ones).

**Comment by ajwerner** [comment]
That approach seems fine. 

Now here’s another wrinkle and what I think works as a solution: what should we do about transactions which have already laid down intents? A pure queueing token bucket rate limiter seems like it could be a source of deadlock.

I was starting to feel like it was overkill that I had built the thing on top of the quotapool abstraction but now I’m feeling pretty good about it. I’m thinking all requests will acquire from the limited when they enter but if they hit a lock they’ll release their acquired tokens back to the limiter and re-acquire after waiting.

**Comment by tbg** [comment]
Note from #47905: the rate limiters we introduce need to also feed into per-tenant metrics. (The prototype that Andrew has already does that, so this is not new, just making it explicitly a requirement).

**Comment by ajwerner** [comment]
Closed by #52331.
